experimental/InvariantTheory/src/InvariantTheory.jl

```diff
@@ -0,0 +1,666 @@
+    R, _ = graded_polynomial_ring(G.field,"x"=>1:n, "y"=>1:n, "z"=>(1:m, 1:m))
+    #R, _ = grade(PolynomialRing(G.field,"x"=>1:n, "y"=>1:n, "z"=>(1:m, 1:m))[1])
+    map1 = hom(vector_ring, R, gens(R)[1:n])
+    new_elem = map1(elem)
+    group_ring = base_ring(X.rep_mat)
+    map2 = hom(group_ring, R, gens(R)[2*n+1:2*n+m^2])
+    new_rep_mat = map_entries(map2, X.rep_mat)
+    new_det = map2(det(G.canonical_representation))
+    f = X.reynolds_v(new_elem, new_rep_mat, new_det, m)
+    reverse_map = hom(R, vector_ring, vcat(gens(vector_ring), [0 for i in 1:n+m^2]))
+    return reverse_map(f)
+end
+
+function reynolds_operator(R::InvariantRing, elem::MPolyRingElem)
+    X = R.representation
+    return reynolds_operator(X, elem)
+end
+
+##########################
+#fast algorithm for invariants of tori
+##########################
+function torus_invariants_fast(W::Vector{Vector{ZZRingElem}}, R::MPolyRing)
+    #no check that length(W[i]) for all i is the same
+    length(W) == ngens(R) || error("number of weights must be equal to the number of generators of the polynomial ring")
+    n = length(W)
+    r = length(W[1])
+    #step 2
+    if length(W[1]) == 1
+        M = zero_matrix(Int, n, 1)
+        for i in 1:n
+            M[i,1] = W[i][1]
+        end
+        C1 = collect(lattice_points(convex_hull(M)))
+    else
+        M = zero_matrix(Int, 2*n, r)
+        for i in 1:n
+            M[i, 1:r] = 2*r*W[i]
+            M[n + i, 1:r] = -2*r*W[i]   
+        end
+        C1 = collect(lattice_points(convex_hull(M)))
+    end
+    
+    #get a Vector{Vector{ZZRingElem}} from Vector{PontVector{ZZRingElem}}
+    C = [[ZZRingElem(0) for i in 1:r] for j in 1:length(C1)]
+    for i in 1:length(C1)
+        for j in 1:r
+            C[i][j] = C1[i][j]
+        end
+    end
+    #step 3
+    S = Vector{Vector{elem_type(R)}}()
+    U = Vector{Vector{elem_type(R)}}()
+    index_0 = 0
+    for point in C
+        if point == [ZZRingElem(0) for i in 1:r] #this may not work CHECK
+            index_0 = findfirst(item -> item == point, C)
+        end
+        c = true
+        for i in 1:n
+            if point == W[i] #check type here TODO
+                push!(S, [gen(R,i)])
+                push!(U, [gen(R,i)])
+                c = false
+                break
+            end
+        end
+        if c == true
+            push!(S, elem_type(R)[])
+            push!(U, elem_type(R)[])
+        end
+    end
+    k = 0
+    while k<4
+        #step 4
+        @label step_4 #not sure if this is needed...
+        j = 0
+        for i in 1:length(U)
+            if length(U[i]) != 0
+                j = i
+                @goto step4_b
+            end
+        end
+        return S[index_0]
+        @label step4_b
+        m = U[j][1]
+        w = C[j] #weight_of_monomial(m, W)
+        #step 5 - 7
+        for i in 1:n
+            u = m*gen(R,i)
+            v = w + W[i]
+            if v in C
+                index = findfirst(item -> item == v, C)
```

Comment:
This could/should be done in one go because `findfirst` returns `nothing` if the element is not in the vector.
```
index = findfirst(isequal(v), C)
if index === nothing
  continue
end
```